File: java/hard/min_rewards/MinRewards.java
package main.java.com.example.minrewards;

import java.util.Arrays;

public class MinRewards {
	// O(n) time | O(n) space, where n is the length of the input array
	public static int minRewards(int[] scores) {
		// Initialize the rewards array with 1 reward for each student.
		int[] rewards = new int[scores.length];
		Arrays.fill(rewards, 1);

		// Forward pass: Ensure rewards increase for increasing scores from left to right, based on the provided rules.
		for (int i = 1; i < scores.length; i++) {
			if (scores[i] > scores[i - 1]) {
				rewards[i] = rewards[i - 1] + 1;
			}
		}

		// Backward pass: Ensure rewards increase for increasing scores from right to left, based on the provided rules.
		for (int i = scores.length - 2; i >= 0; i--) {
			if (scores[i] > scores[i + 1]) {
				rewards[i] = Math.max(rewards[i], rewards[i + 1] + 1);
			}
		}

		// Calculate the total sum of rewards.
		int totalRewards = 0;
		for (int reward : rewards) {
			totalRewards += reward;
		}

		// Return the total sum of rewards.
		return totalRewards;
	}
}

File: python/hard/min_rewards/min_rewards.py
# O(n) time | O(n) space, where n is the length of the input array
# Function to return the minimum number of rewards to be given out to students based on their scores in an array.
def minRewards(scores):
    # Initialize the rewards array with 1 reward for each student.
    rewards = [1 for _ in scores]

    # Forward pass: Ensure rewards increase for increasing scores from left to right, based on the provided rules.
    for i in range(1, len(scores)):
        if scores[i] > scores[i - 1]:
            rewards[i] = rewards[i - 1] + 1

    # Backward pass: Ensure rewards increase for increasing scores from right to left, based on the provided rules.
    for i in reversed(range(len(scores) - 1)):
        if scores[i] > scores[i + 1]:
            rewards[i] = max(rewards[i], rewards[i + 1] + 1)

    # Return the total sum of rewards.
    return sum(rewards)
